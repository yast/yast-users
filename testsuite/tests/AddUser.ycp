/**
 * File:
 *  AddGroup.ycp
 *
 * Module:
 *  Users configurator
 *
 * Summary:
 *  Saving group tests.
 *
 * Authors:
 *  Jiri Suchomel <jsuchome@suse.cz>
 *
 */

{
    // testedfiles: Users.pm UsersLDAP.pm UsersCache.pm

    import "Directory";
    import "Users";
    import "UsersCache";
    import "UsersLDAP";
    import "Mode";
    import "Progress";
    import "Report";

    string tmpdir	= Directory::tmpdir;
    foreach (string file, [ "passwd", "group", "shadow" ], {
	string cmd	= sformat ("/bin/cp ./%1 %2/", file, tmpdir);
	SCR::Execute (.target.bash_output, cmd);
    });
    Users::SetBaseDirectory (tmpdir);
    Users::ReadLocal ();

    map READ = $[
	"etc" : $[
	    "fstab": [],
	    "cryptotab": [],
	    "default": $[
		"useradd": $[
		    "home":	"/home",
		    "groups":	"audio,video",
		    "inactive":	nil,
		    "expire":	nil,
		    "shell":	nil,
		    "group":	100
		]
	    ],
	],
	"target": $[
	    "stat"	: $[],
	    "size"	: -1,
	    "tmpdir"	: "/tmp/YaST"
	],
	"product": $[
		"features": $[
		    "USE_DESKTOP_SCHEDULER"	: "no",
		    "IO_SCHEDULER"		: "",
		    "ENABLE_AUTOLOGIN"		: "false",
		    "UI_MODE"			: "simple",
		    "EVMS_CONFIG"		: "no",
		    "INCOMPLETE_TRANSLATION_TRESHOLD"	: "99",
		]
	],
    ];
    map WRITE = $[];
    map EXEC = $[
	"passwd" : $[
	    "init"	: true
	],
	"target" : $[
	    "bash"		: 0,
	    "bash_output"	: $[]
	],
    ];

    import "Testsuite";

    Testsuite::Dump ("==========================================================");
    Mode::SetTest ("test");

    Testsuite::Test (``(Users::Read ()), [READ, WRITE, EXEC], 0);

    Testsuite::Dump (sformat ("---- current user:\n %1", Users::GetCurrentUser ()));
    Testsuite::Test (``(Users::AddUser ($[])), [READ, WRITE, EXEC], 0);

    Testsuite::Dump (sformat ("---- current user (empty add):\n %1", Users::GetCurrentUser ()));
    map<string,any> user = $[
	"uidNumber"	: 501,
	"uid"		:"aaa",
	"gidNumber"	: 100,
	"groupname"	: "users",
	"grouplist"	: $[ "audio" : 1 ],
	"homeDirectory"	:"/local/home/aaa",
	"userPassword"	:"qqqqq",
	"type"		:"local",
    ];

    Testsuite::Test (``(Users::AddUser (user)), [READ, WRITE, EXEC], 0);

    Testsuite::Dump (sformat ("---- current user (after rich add):\n %1", Users::GetCurrentUser ()));

    EXEC ["target", "bash_output", "stdout"] = "aaa";

    string error = (string) Testsuite::Test(``(Users::CheckUser ($[])), [READ,WRITE,EXEC], 0);

    Users::SelectGroup (100);
    Testsuite::Dump (sformat ("---- current group (before user commit):\n %1",
	Users::GetCurrentGroup ()));

    Testsuite::Dump (sformat ("---- check user after add:\n %1", error));

    Testsuite::Dump ("---- commit user:");

    Testsuite::Test (``(Users::CommitUser ()), [READ, WRITE, EXEC], 0);

    Users::SelectGroup (100);
    Testsuite::Dump (sformat ("---- current group (afer user commit):\n %1",
	Users::GetCurrentGroup ()));

    Testsuite::Dump ("=================== no password ==========================");

    Testsuite::Test (``(Users::AddUser ($[])), [READ, WRITE, EXEC], 0);
    Testsuite::Test (``(Users::AddUser ($[ "uid": "hhh"])), [READ, WRITE, EXEC], 0);

    Testsuite::Dump (sformat ("---- current user (minimal add, used default values):\n %1", Users::GetCurrentUser ()));

    EXEC ["target", "bash_output", "stdout"] = "hhh";

    error = (string) Testsuite::Test(``(Users::CheckUser ($[])), [READ,WRITE,EXEC], 0);

    Testsuite::Dump (sformat ("---- check user after add:\n %1", error));

    Testsuite::Dump ("==================== username conflict ========================");

    Testsuite::Test (``(Users::AddUser ($[])), [READ, WRITE, EXEC], 0);
    Testsuite::Test (``(Users::AddUser ($[ "uid": "root", "userPassword": "qqqqq"])),
	[READ, WRITE, EXEC], 0);

    Testsuite::Dump (sformat ("---- current user (just added):\n %1", Users::GetCurrentUser ()));
    EXEC ["target", "bash_output", "stdout"] = "root";

    error = (string) Testsuite::Test(``(Users::CheckUser ($[])), [READ,WRITE,EXEC], 0);

    Testsuite::Dump (sformat ("---- check user after add:\n %1", error));

    Testsuite::Dump ("==================== uidNumber problems ===================");

    Testsuite::Test (``(Users::AddUser ($[])), [READ, WRITE, EXEC], 0);
    Testsuite::Test (``(Users::AddUser ($[
	"uid"		: "rrr",
	"uidNumber"	: 5,
	"type"		: "local",
	"userPassword"	: "qqqqq"
	])),
	[READ, WRITE, EXEC], 0);

    Testsuite::Dump (sformat ("---- current user (just added):\n %1", Users::GetCurrentUser ()));
    EXEC ["target", "bash_output", "stdout"] = "rrr";

    error = (string) Testsuite::Test(``(Users::CheckUser ($[])), [READ,WRITE,EXEC], 0);

    Testsuite::Dump (sformat ("---- check user after add:\n %1", error));

    Testsuite::Dump ("==================== uid content problems ===================");

    Testsuite::Test (``(UsersLDAP::SetUserBase ("dc=suse,dc=cz")), [], 0);
    Testsuite::Test (``(Users::AddUser ($[])), [READ, WRITE, EXEC], 0);
    Testsuite::Test (``(Users::AddUser ($[
	"uid"		: "iii$",
	"type"		: "local",
	"userPassword"	: "qqqqq"
	])),
	[READ, WRITE, EXEC], 0);

    Testsuite::Dump (sformat ("---- current user (just added):\n %1", Users::GetCurrentUser ()));
    EXEC ["target", "bash_output", "stdout"] = "";

    error = (string) Testsuite::Test(``(Users::CheckUser ($[])), [READ,WRITE,EXEC], 0);

    Testsuite::Dump (sformat ("---- check user after add:\n %1", error));

    Testsuite::Dump ("==================== uid for LDAP may contain '$' ==============");

    Testsuite::Test (``(Users::AddUser ($[])), [READ, WRITE, EXEC], 0);
    Testsuite::Test (``(Users::AddUser ($[
	"uid"		: "iii$",
	"type"		: "ldap",
	"homeDirectory"	: "/home/ldap/iii",
	"userPassword"	: "qqqqq"
	])),
	[READ, WRITE, EXEC], 0);

    Testsuite::Dump (sformat ("---- current user (just added):\n %1", Users::GetCurrentUser ()));
    EXEC ["target", "bash_output", "stdout"] = "iii";

    error = (string) Testsuite::Test(``(Users::CheckUser ($[])), [READ,WRITE,EXEC], 0);

    Testsuite::Dump (sformat ("---- check user after add:\n %1", error));

    Testsuite::Dump ("==========================================================");

    Testsuite::Dump ("============== duplicated UID number =====================");

    Testsuite::Test (``(Users::AddUser ($[])), [READ, WRITE, EXEC], 0);
    Testsuite::Test (``(Users::AddUser ($[ "uid": "admin", "uidNumber": 0,
	"userPassword": "qqqqq", "type" : "system"])), [READ, WRITE, EXEC], 0);

    EXEC ["target", "bash_output", "stdout"] = "admin";

    error = (string) Testsuite::Test(``(Users::CheckUser ($[])), [READ,WRITE,EXEC], 0);
}
