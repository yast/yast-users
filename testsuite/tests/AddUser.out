Dump	==========================================================
Read	.etc.default.useradd."expire" nil
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/home"
Read	.etc.default.useradd."inactive" nil
Read	.etc.default.useradd."shell" nil
Read	.etc.default.useradd."skel" 0
Read	.etc.default.useradd."umask" 0
Read	.target.string "/etc/shells" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.etc.nsswitch_conf.passwd 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $[] true
Return	
Dump	---- current user:
Dump	 $[]
Return	
Dump	---- current user (empty add):
Dump	 $["chown_home":true, "cn":"", "create_home":true, "gidNumber":"100", "grouplist":$["audio":"1", "video":"1"], "groupname":"users", "loginShell":"", "plugins":[], "shadowExpire":"", "shadowFlag":"", "shadowInactive":"", "shadowLastChange":"0", "shadowMax":"99999", "shadowMin":"0", "shadowWarning":"7", "type":"local", "uidNumber":"1000", "userPassword":nil, "what":"add_user"]
Return	
Dump	---- current user (after rich add):
Dump	 $["chown_home":true, "cn":"", "create_home":true, "encrypted":true, "gidNumber":"100", "grouplist":$["audio":"1"], "groupname":"users", "homeDirectory":"/local/home/aaa", "loginShell":"", "plugins":[], "shadowExpire":"", "shadowFlag":"", "shadowInactive":"", "shadowLastChange":"0", "shadowMax":"99999", "shadowMin":"0", "shadowWarning":"7", "text_userpassword":"qqqqq", "type":"local", "uid":"aaa", "uidNumber":"501", "userPassword":"crypted_qqqqq", "what":"add_user"]
Execute	.target.bash_output "echo 'aaa' | grep '^[[:alpha:]_][[:alnum:]_.-]*[[:alnum:]_.$-]\\?$'" $["LANG":"C"] $["stdout":"aaa"]
Return	
Dump	---- current group (before user commit):
Dump	 $["cn":"users", "gidNumber":"100", "more_users":$["hh":"1", "ii":"1"], "type":"local", "userPassword":"x", "userlist":$[]]
Dump	---- check user after add:
Dump	 
Dump	---- commit user:
Return	true
Dump	---- current group (afer user commit):
Dump	 $["cn":"users", "gidNumber":"100", "more_users":$["aaa":"1", "hh":"1", "ii":"1"], "org_cn":"users", "org_gidNumber":"100", "type":"local", "userPassword":"x", "userlist":$[], "what":"user_change_default"]
Dump	=================== no password ==========================
Return	
Return	
Dump	---- current user (minimal add, used default values):
Dump	 $["chown_home":true, "cn":"", "create_home":true, "gidNumber":"100", "grouplist":$["audio":"1", "video":"1"], "groupname":"users", "homeDirectory":"/home/hhh", "loginShell":"", "plugins":[], "shadowExpire":"", "shadowFlag":"", "shadowInactive":"", "shadowLastChange":"0", "shadowMax":"99999", "shadowMin":"0", "shadowWarning":"7", "type":"local", "uid":"hhh", "uidNumber":"1000", "userPassword":nil, "what":"add_user"]
Execute	.target.bash_output "echo 'hhh' | grep '^[[:alpha:]_][[:alnum:]_.-]*[[:alnum:]_.$-]\\?$'" $["LANG":"C"] $["stdout":"hhh"]
Return	No password entered.
Dump	---- check user after add:
Dump	 No password entered.
Dump	Try again.
Dump	==================== username conflict ========================
Return	
Return	
Dump	---- current user (just added):
Dump	 $["chown_home":true, "cn":"", "create_home":true, "encrypted":true, "gidNumber":"100", "grouplist":$["audio":"1", "video":"1"], "groupname":"users", "homeDirectory":"/home/root", "loginShell":"", "plugins":[], "shadowExpire":"", "shadowFlag":"", "shadowInactive":"", "shadowLastChange":"0", "shadowMax":"99999", "shadowMin":"0", "shadowWarning":"7", "text_userpassword":"qqqqq", "type":"local", "uid":"root", "uidNumber":"1000", "userPassword":"crypted_qqqqq", "what":"add_user"]
Execute	.target.bash_output "echo 'root' | grep '^[[:alpha:]_][[:alnum:]_.-]*[[:alnum:]_.$-]\\?$'" $["LANG":"C"] $["stdout":"root"]
Return	There is a conflict between the entered
Dump	---- check user after add:
Dump	 There is a conflict between the entered
Dump	username and an existing username. 
Dump	Try another one.
Dump	==================== uidNumber problems ===================
Return	
Return	
Dump	---- current user (just added):
Dump	 $["chown_home":true, "cn":"", "create_home":true, "encrypted":true, "gidNumber":"100", "grouplist":$["audio":"1", "video":"1"], "groupname":"users", "homeDirectory":"/home/rrr", "loginShell":"", "plugins":[], "shadowExpire":"", "shadowFlag":"", "shadowInactive":"", "shadowLastChange":"0", "shadowMax":"99999", "shadowMin":"0", "shadowWarning":"7", "text_userpassword":"qqqqq", "type":"local", "uid":"rrr", "uidNumber":"5", "userPassword":"crypted_qqqqq", "what":"add_user"]
Return	The selected user ID is not allowed.
Dump	---- check user after add:
Dump	 The selected user ID is not allowed.
Dump	Select a valid integer between 1000 and 60000.
Dump	==================== uid content problems ===================
Return	nil
Return	
Return	
Dump	---- current user (just added):
Dump	 $["chown_home":true, "cn":"", "create_home":true, "encrypted":true, "gidNumber":"100", "grouplist":$["audio":"1", "video":"1"], "groupname":"users", "homeDirectory":"/home/iii$", "loginShell":"", "plugins":[], "shadowExpire":"", "shadowFlag":"", "shadowInactive":"", "shadowLastChange":"0", "shadowMax":"99999", "shadowMin":"0", "shadowWarning":"7", "text_userpassword":"qqqqq", "type":"local", "uid":"iii$", "uidNumber":"1000", "userPassword":"crypted_qqqqq", "what":"add_user"]
Execute	.target.bash_output "echo 'iii$' | grep '^[[:alpha:]_][[:alnum:]_.-]*[[:alnum:]_.$-]\\?$'" $["LANG":"C"] $["stdout":""]
Return	The username may contain only
Dump	---- check user after add:
Dump	 The username may contain only
Dump	letters, digits, "-", ".", and "_"
Dump	and must begin with a letter or "_".
Dump	Try again.
Dump	==================== uid for LDAP may contain '$' ==============
Return	
Return	
Dump	---- current user (just added):
Dump	 $["chown_home":true, "cn":"", "create_home":true, "dn":"uid=iii$,dc=suse,dc=cz", "encrypted":true, "gidNumber":"100", "grouplist":$["audio":"1", "video":"1"], "groupname":"users", "homeDirectory":"/home/ldap/iii", "loginShell":"", "plugins":[], "shadowExpire":"", "shadowFlag":"", "shadowInactive":"", "shadowLastChange":"0", "shadowMax":"99999", "shadowMin":"0", "shadowWarning":"7", "text_userpassword":"qqqqq", "type":"ldap", "uid":"iii$", "uidNumber":"1000", "userPassword":"crypted_qqqqq", "what":"add_user"]
Execute	.target.bash_output "echo 'iii' | grep '^[[:alpha:]_][[:alnum:]_.-]*[[:alnum:]_.$-]\\?$'" $["LANG":"C"] $["stdout":"iii"]
Return	
Dump	---- check user after add:
Dump	 
Dump	==========================================================
Dump	============== duplicated UID number =====================
Return	
Return	
Execute	.target.bash_output "echo 'admin' | grep '^[[:alpha:]_][[:alnum:]_.-]*[[:alnum:]_.$-]\\?$'" $["LANG":"C"] $["stdout":"admin"]
Return	
